### ISSUE TITLE
Decoding Natural overflows in GHC 9

### ISSUE THREAD

**Issue by brandon-leapyear** [haskell/aeson#858]
Running with `stack ghci --resolver lts-18.0 --compiler ghc-9.0.1 --package aeson`:
```hs
import Data.Aeson
import Numeric.Natural
:set -XTypeApplications -XOverloadedStrings

-- Just 18446744073709551615
decode @Natural "18446744073709551615"

-- Just 0
decode @Natural "18446744073709551616"
```

**Comment by brandon-leapyear** [comment]
huh. I cloned `aeson` locally and it seems like it's working on `master`? Out of curiosity, which PR fixed this?

If `master` is working, when can a Hackage release be cut?

**Comment by phadej** [comment]
This looks like https://gitlab.haskell.org/ghc/ghc/-/issues/19411, i.e. GHC is just broken.

**Comment by brandon-leapyear** [comment]
I see. That seems possible, although I do have it working in a local `aeson` clone.
1. Clone `haskell/aeson`
2. Edit `stack-nightly.yaml` to use `nightly-2021-07-12`
3. `stack ghci --stack-yaml stack-nightly.yaml`
4. Verify GHC is using 9.0.1
5. `decode @Natural "18446744073709551616"` returns 18446744073709551616

**Comment by phadej** [comment]
Yes, `-O0` turns of the optimizations, and broken rule doesn't fire.

**Comment by brandon-leapyear** [comment]
Fixed in GHC 9.0.2
